src/main/java/com/vesperin/cue/segment/BlockSegmentationGraph.java

```diff
@@ -0,0 +1,282 @@
+   */
+  private static final int VISIT_COLOR_WHITE = 1;
+
+  /**
+   * Color used to mark nodes as they are first visited in DFS order
+   */
+  private static final int VISIT_COLOR_GREY = 2;
+
+  /**
+   * Color used to mark nodes after descendants are completely visited
+   */
+  private static final int VISIT_COLOR_BLACK = 4;
+
+  /**
+   * Constructs an empty block segmentation graph (DAG).
+   */
+  BlockSegmentationGraph(){
+    this(new EdgeFactoryImpl<>());
+  }
+  /**
+   * Constructs an empty block segmentation graph (DAG) with a predefined
+   * edge factory object.
+   *
+   * @param edgeFactory application specific object for creating edges.
+   * @throws NullPointerException if providing a null edge factory
+   */
+  private BlockSegmentationGraph(EdgeFactory<Segment, Edge<Segment>> edgeFactory) {
+    super(edgeFactory);
+  }
+
+  @Override public boolean hasCycle() {
+    return !findCycles(this).isEmpty();
+  }
+
+  @Override public Segment segmentBy(String label) {
+    return vertexSet().stream()
+      .filter(v -> Objects.equals(v.label(), label))
+      .findFirst().orElse(null);
+  }
+
+  private static List<Edge<Segment>> findCycles(DirectedAcyclicGraph<Segment, Edge<Segment>> graph) {
+
+    final List<Edge<Segment>> cycleEdges = new ArrayList<>();
+    for(Segment each : graph.vertexSet()){
+      // Mark all vertices as white
+      each.setMarkState(VISIT_COLOR_WHITE);
+    }
+
+    for(Segment each : graph.vertexSet()){
+      visit(each, graph, cycleEdges);
+    }
+
+    return cycleEdges;
+  }
+
+
+
+  private static void visit(Segment v, DirectedAcyclicGraph<Segment, Edge<Segment>> graph,
+          List<Edge<Segment>> cycleEdges) {
+
+    v.setMarkState(VISIT_COLOR_GREY);
+
+    for(Edge<Segment> each : graph.outgoingEdgesOf(v)){
+      Segment u = each.to();
+      if(u.markState() == VISIT_COLOR_GREY) {
+        cycleEdges.add(each);
+      } else if (u.markState() == VISIT_COLOR_WHITE){
+        visit(u, graph, cycleEdges);
+      }
+    }
+
+    v.setMarkState(VISIT_COLOR_BLACK);
+  }
+
+  @Override public Segment segmentAt(int n) {
+    return Iterables.get(vertexSet(), n);
+  }
+
+  @Override public Set<Location> irrelevantSet(int capacity) {
+    if(capacity <= 3) return ImmutableSet.of();
+
+    final List<Segment> blackList = generateBlackList(this, capacity);
+
+    return blackList.stream()
+      .map(foldable -> Locations.locate(foldable.data()))
+      .collect(toImmutableSet());
+  }
+
+  /**
+   * Creates a collector that transforms a mutable set into an immutable one.
+   *
+   * @param <T> the type parameter.
+   * @return a new collector object.
+   */
+  private static <T> Collector<T, ?, Set<T>> toImmutableSet() {
+    return Collectors.collectingAndThen(
+      Collectors.toSet(),
+      Collections::unmodifiableSet
+    );
+  }
+
+  private static List<Segment> generateBlackList(SegmentationGraph graph, int capacity) {
+    final List<Segment> allSegments = Lists.newLinkedList(
```

Comment:
should we really exclude the root? Put your answer in the context of multistaging.